front/lib/api/assistant/generation.ts

```diff
@@ -235,12 +238,15 @@ export async function renderConversationForModel({
     },
     "[ASSISTANT_TRACE] Genration message token counts for model conversation rendering"
   );
+  while (selected.length > 0 && selected[0].message.role === "agent") {
+    selected.shift();
```

Comment:
```suggestion
    tokensCount -= messageCountRes[messages.length - selected.length]
    selected.shift();
```
would this suggestion work? it would allow not having to make all the other modifications to messages and adding a tokencount to each message, limiting the PR to 4 lines?

Reply:
Really think it's worth doing

Reply:
Looks like it would work but it makes it quite hard to follow and more error prone IMHO.
I am in favor of keeping it as is.

Reply:
Ok, we can disagree on that. IMHO it's easier to follow and less error prone: 
1. simpler type for the array (messages VS {messages, tokenUsed}), 
2. the change is concentrated in one place VS 5 changes at different places, with a map & a throw

Reply:
This being said, given you already have 2 approvals, I will not block the merge, do as you see fit :+1:

Reply:
FWIW I agree that @philipperolet's proposal is simpler